### PR TITLE
add extraction of binaries from RPM archives

### PR DESCRIPTION
**Pull request by itraviv**
* detecting ELF members
* resolving links in rpms and adding them as members using the ino

### CODE REVIEW

File: rpmfile/__init__.py [hunk 1]
```diff
@@ -1,44 +1,82 @@
...
 from __future__ import print_function, unicode_literals, absolute_import
-from .headers import get_headers
-import sys
-import io
+
 import gzip
+import io
+import sys
+
+from .headers import get_headers
```
Comment (yammesicka) [inline rpmfile/__init__.py:7]: Why relative?

File: rpmfile/__init__.py [hunk 2]
```diff
@@ -1,44 +1,82 @@
...
         self.initial_offset = initial_offset
         self._isdir = isdir
+        self._ino = ino
+        self._is_elf = is_elf
+        self.link_target = None
+
+    @property
+    def file_start(self):
+        if self.link_target is not None:
+            assert self.link_target != self
```
Comment (yammesicka) [inline rpmfile/__init__.py]: What does it mean?

File: rpmfile/__init__.py [hunk 3]
```diff
@@ -108,13 +152,31 @@ def __exit__(self, *excinfo):
 
     _members = None
 
+    @staticmethod
+    def _resolve_links(ino_map):
+        for _, _ino_members in ino_map.items():
```
Comment (yammesicka) [inline rpmfile/__init__.py]: You use it inside the for, so `ino_members`

File: rpmfile/__init__.py [hunk 4]
```diff
@@ -108,13 +152,31 @@ def __exit__(self, *excinfo):
...
+            _links_or_dirs = []
+            target_member = None
+            for member in _ino_members:
+                if member.size > 0 and not member.isdir:
+                    target_member = member
+                else:
+                    _links_or_dirs.append(member)
+            if target_member is None:
```
Comment (yammesicka) [inline rpmfile/__init__.py]: I would prefer `is not None` and indentation, but your choice

File: rpmfile/__init__.py [hunk 5]
```diff
@@ -108,13 +152,31 @@ def __exit__(self, *excinfo):
 
     _members = None
 
+    @staticmethod
+    def _resolve_links(ino_map):
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Too long, maybe to split to methods?

File: rpmfile/__init__.py [hunk 6]
```diff
@@ -108,13 +152,31 @@ def __exit__(self, *excinfo):
...
-        '''
+        """
         Return the members of the archive as a list of RPMInfo objects. The
         list has the same order as the members in the archive.
-        '''
+        """
         if self._members is None:
-            self._members = _members = []
+            _members = []
+            from collections import defaultdict
```
Comment (yammesicka) [inline rpmfile/__init__.py]: ?

File: rpmfile/__init__.py [hunk 7]
```diff
@@ -124,41 +186,49 @@ def getmembers(self):
...
-
-                    if not member.isdir:
-                        _members.append(member)
+                    _ino_map[member.ino].append(member)
+                    _members.append(member)
 
                 magic = g.read(2)
-            return _members
+            self._resolve_links(_ino_map)
+            self._members = list(filter(lambda x: not x.isdir, _members))
```
Comment (yammesicka) [inline rpmfile/__init__.py]: List comprehension is better here

File: rpmfile/__init__.py [hunk 8]
```diff
@@ -124,41 +186,49 @@ def getmembers(self):
...
 
     _data_file = None
 
+    def get_binaries(self):
+        """
+        Get a list all members that were identified as ELF
+        :return: List of RPMInfo, each represent a member identified as ELF
+        """
+
+        return list(filter(lambda m: m.is_elf, self.getmembers()))
```
Comment (yammesicka) [inline rpmfile/__init__.py]: `Type.is_elf` instead of lambda?

File: rpmfile/__init__.py [hunk 9]
```diff
@@ -124,41 +186,49 @@ def getmembers(self):
...
-        '''
+        """
         if not isinstance(member, RPMInfo):
             member = self.getmember(member)
         return _SubFile(self.data_file, member.file_start, member.size)
 
     _data_file = None
 
+    def get_binaries(self):
+        """
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Get should start here, see PEP 257

File: rpmfile/__init__.py [hunk 10]
```diff
@@ -124,41 +186,49 @@ def getmembers(self):
...
+
         for m in members[::-1]:
             if m.name == name:
                 return m
-
         raise KeyError("member %s could not be found" % name)
 
     def extractfile(self, member):
-        '''
+        """
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Extract should be here, PEP 257

File: rpmfile/__init__.py [hunk 11]
```diff
@@ -108,13 +152,31 @@ def __exit__(self, *excinfo):
 
     _members = None
 
+    @staticmethod
+    def _resolve_links(ino_map):
+        for _, _ino_members in ino_map.items():
+            _links_or_dirs = []
+            target_member = None
+            for member in _ino_members:
+                if member.size > 0 and not member.isdir:
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Possible bug if this branch gets True more than once. if not possible, add assertion

File: rpmfile/__init__.py [hunk 12]
```diff
@@ -1,44 +1,82 @@
...
+        self._file_start = file_start
+        self._size = file_size
         self.initial_offset = initial_offset
         self._isdir = isdir
+        self._ino = ino
+        self._is_elf = is_elf
+        self.link_target = None
+
+    @property
+    def file_start(self):
```
Comment (yammesicka) [inline rpmfile/__init__.py:54]: Maybe you should export all these to a single function

File: rpmfile/__init__.py (hunk 1 above)
Comment (itraviv) [reply rpmfile/__init__.py:7]: was originally in the sources

File: rpmfile/__init__.py (hunk 9 above)
Comment (itraviv) [reply rpmfile/__init__.py]: nice to know. thanks

File: rpmfile/__init__.py (hunk 2 above)
Comment (itraviv) [reply rpmfile/__init__.py]: it should avoid links to self (circular dependency), added assertion message

File: rpmfile/__init__.py (hunk 5 above)
Comment (itraviv) [reply rpmfile/__init__.py]: agreed. splitted

File: rpmfile/__init__.py (hunk 12 above)
Comment (itraviv) [reply rpmfile/__init__.py:54]: belongs to the base open source lib, used across the lib.

File: .gitignore [hunk 13]
```diff
@@ -1,3 +1,5 @@
 *.pyc
 *.pyor
 *.egg-info/
+*venv/**
+.idea/**
```
Comment (yammesicka) [inline .gitignore]: Newline

File: rpmfile/__init__.py [hunk 14]
```diff
@@ -1,44 +1,85 @@
-
 from __future__ import print_function, unicode_literals, absolute_import
-from .headers import get_headers
-import sys
-import io
+
```
Comment (yammesicka) [inline rpmfile/__init__.py]: `__future__` is a standard library, don't split from the rest

File: rpmfile/__init__.py [hunk 15]
```diff
@@ -1,44 +1,85 @@
...
         self.initial_offset = initial_offset
         self._isdir = isdir
+        self._ino = ino
+        self._is_elf = is_elf
+        self.link_target = None
+
+    @property
+    def file_start(self):
+        if self.link_target is not None:
+            assert (
```
Comment (yammesicka) [inline rpmfile/__init__.py:47]: Are you sure this is an assertion?

File: rpmfile/__init__.py (hunk 12 above)
Comment (yammesicka) [reply rpmfile/__init__.py:54]: *reopen*

File: rpmfile/__init__.py [hunk 16]
```diff
@@ -55,36 +96,42 @@ def _read_new(cls, fileobj, magic=None):
...
+        is_elf = fileobj.read(4) == b'\x7fELF'
         file_size = int(d[6], 16)
-        fileobj.seek(file_size, 1)
+        fileobj.seek(file_size - 4, 1)
         fileobj.seek(pad(fileobj), 1)
         nlink = int(d[4], 16)
+        ino = d[0] + d[1]
         isdir = nlink == 2 and file_size == 0
-        return cls(name, file_start, file_size, initial_offset, isdir)
+        return cls(name, file_start, file_size, initial_offset, isdir, ino,
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Split lines, hanging indent isn't look so well

File: rpmfile/__init__.py (hunk 6 above)
Comment (yammesicka) [reply rpmfile/__init__.py]: *reopen*

File: rpmfile/__init__.py [hunk 17]
```diff
@@ -108,13 +155,41 @@ def __exit__(self, *excinfo):
 
     _members = None
 
+    @staticmethod
+    def _assign_link_to_members(links_or_dirs, link_target):
+        if link_target is None:
+            return
```
Comment (yammesicka) [inline rpmfile/__init__.py:164]: Split the logic with a linebreak

File: rpmfile/__init__.py [hunk 18]
```diff
@@ -108,13 +155,41 @@ def __exit__(self, *excinfo):
...
+        for member in links_or_dirs:
+            member.link_target = link_target
+            member.isdir = False
+
+    @classmethod
+    def _resolve_links_for_same_ino_members(cls, ino_members):
+        """All members in ino_members have the same inode, thus
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Please read PEP 257

File: rpmfile/__init__.py [hunk 19]
```diff
@@ -108,13 +155,41 @@ def __exit__(self, *excinfo):
...
+        referring to the same file"""
+        _links_or_dirs = []
+        target_member = None
+        for member in ino_members:
+            if member.size > 0 and not member.isdir:
+                assert target_member is None, "More than one target found"
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Shouldn't be an exception?

File: rpmfile/__init__.py [hunk 20]
```diff
@@ -124,41 +199,46 @@ def getmembers(self):
...
         return _SubFile(self.data_file, member.file_start, member.size)
 
     _data_file = None
 
+    def get_binaries(self):
+        """Get a list all members that were identified as ELF
+        :return: List of RPMInfo, each represent a member identified as ELF
+        """
+
+        return list(filter(lambda m: m.is_elf, self.getmembers()))
```
Comment (yammesicka) [inline rpmfile/__init__.py]: Looks to me like `[m.is_elf for m in self.getmembers() if m.is_elf]`

File: rpmfile/__init__.py [hunk 21]
```diff
@@ -124,41 +199,46 @@ def getmembers(self):
...
-        '''
+        """
         if not isinstance(member, RPMInfo):
             member = self.getmember(member)
         return _SubFile(self.data_file, member.file_start, member.size)
 
     _data_file = None
 
+    def get_binaries(self):
+        """Get a list all members that were identified as ELF
```
Comment (yammesicka) [inline rpmfile/__init__.py:247]: Linebreak

**Review by yammesicka — CHANGES_REQUESTED** [review]

**Review by yammesicka — APPROVED** [review]

**Review by yammesicka — APPROVED** [review]

**Pull request merged**